### DISCUSSION TITLE
How to pass metadata to all tasks submitted by the same client?

### DISCUSSION THREAD

**Discussion by averri** [dask/dask#9409] [Q&A]
I would like to pass any metadata to all tasks associated with a particular client. The main objective is to add a unique ID that is part of a large context, this unique ID must be printed in the log of the workers for correlation. I have tried using annotations, but it's not clear how to extract the annotations from the worker's code.

A simple example:

```python
import time
from concurrent.futures import ThreadPoolExecutor
from uuid import uuid4

from dask import annotate
from distributed import Client


def task(tid: int):
    # How to get the request_id from here?
    print(f"Task {tid} started.")
    time.sleep(2)
    print(f"Task {tid} finished.")


def submit_task(tid: int):
    client = Client("localhost:8888")
    # I want this request_id to be available to the submitted task.
    request_id = str(uuid4())
    with annotate(request_id=request_id):
        print(f"Submitting tid={tid}...")
        client.submit(task, tid, pure=False)


if __name__ == '__main__':
    with ThreadPoolExecutor() as exe:
        for i in range(4):
            exe.submit(submit_task, I)
```

In the example above, 4 different threads submit a  task to Dask. Each submission has a unique `request_id` that should be available for the tasks that belong to the same client that submitted the tasks. The `request_id` **must not be part of the task signature**, because it does not belong to the domain model, this is a cross-cutting aspect that is used only for logging.  

How to extract the `request_id` from each task? If annotations are not for this purpose, how to achieve this goal?

**Comment by pavithraes** [comment]
@averri Thanks for your question, we prefer [Discourse](https://dask.discourse.group/) for usage questions, would you mind moving this discussion there?

**Comment by averri** [comment]
Hi @pavithraes, I have found a way to get the annotations from the tasks' code, it's a bit cumbersome, it requires executing a code in the scheduler. But it seems there is a bug with the implementation. I have opened this issue: https://github.com/dask/distributed/issues/6930

I also have a suggestion to improve Dask, making the annotations available in the tasks' code, in a similar way the `key` is available using :
```python
from distributed.worker import thread_state
# Get the tasks' key.
thread_state.key
# I also would like to have:
thread_state.annotations
```

This is the issue for the feature suggestion: https://github.com/dask/dask/issues/9414
